extensions/amp-a4a/0.1/real-time-config-manager.js

```diff
@@ -0,0 +1,163 @@
+    promiseArray.push(sendRtcCallout_(
+        url, rtcStartTime, win, rtcConfig['timeoutMillis'], urlToCalloutMap[url]));
+  }
+  return Promise.all(promiseArray);
+}
+
+function sendRtcCallout_(url, rtcStartTime, win, timeoutMillis, callout) {
+  /**
+   * Note: Timeout is enforced by timerFor, not the value of
+   *   rtcTime. There are situations where rtcTime could thus
+   *   end up being greater than timeoutMillis.
+   */
+  return Services.timerFor(win).timeoutPromise(
+      timeoutMillis,
+      Services.xhrFor(win).fetchJson(
+          url, {credentials: 'include'}).then(res => {
+            return res.text().then(text => {
+              rtcTime = Date.now() - rtcStartTime;
+              // An empty text response is allowed, not an error.
+              if (!text) {
+                return {rtcTime, callout};
+              }
+              const rtcResponse = tryParseJson(text);
+              return rtcResponse ? {rtcResponse, rtcTime, callout} :
+              {rtcTime, callout, error: 'Unparsable JSON'};
+            });
+          })).catch(error => {
+            return {error, rtcTime: Date.now() - rtcStartTime, callout};
+          });
+}
+
+/**
+ * Attempts to parse the publisher-defined RTC config off the amp-ad
+ * element, then validates that the rtcConfig exists, and contains
+ * an entry for either vendor URLs, or publisher-defined URLs. If the
+ * config contains an entry for timeoutMillis, validates that it is a
+ * number, or converts to a number if number-like, otherwise overwrites
+ * with the default.
+ * @return {!boolean}
+ */
+function validateRtcConfig(element) {
+  const defaultTimeoutMillis = 1000;
+  const rtcConfig = tryParseJson(
+      element.getAttribute('prerequest-callouts'));
+  if (!rtcConfig) {
+    return false;
+  }
+  try {
+    user().assert(rtcConfig['vendors'] || rtcConfig['urls'],
+                  'RTC Config must specify vendors or urls');
+    user().assert(!rtcConfig['vendors'] || isObject(rtcConfig['vendors']),
+                  'RTC invalid vendors');
+    user().assert(!rtcConfig['urls'] || isArray(rtcConfig['urls']),
+                  'RTC invalid urls');
+  } catch (err) {
+    return false;
+  }
+  let timeout = rtcConfig['timeoutMillis'];
+  if (timeout) {
+    if (!Number.isInteger(timeout) || timeout >= defaultTimeoutMillis || timeout < 0) {
+      timeout = defaultTimeoutMillis;
+      user().warn(TAG, `Invalid RTC timeout: ${timeout}ms, ` +
+                  `using default timeout ${timeoutMillis}ms`);
+    }
+  }
+  rtcConfig['timeoutMillis'] = timeout;
+  return rtcConfig;
+}
+
+function maybeInflatePublisherUrls(rtcConfig, urlToCalloutMap, customMacros, ampDoc) {
+  if (!rtcConfig['urls']) {
+    return;
+  }
+  let url;
+  let remaining;
+  for (let i in rtcConfig['urls']) {
+    url = rtcConfig['urls'][i];
+    maybeInflateAndAddUrl(url, customMacros, ampDoc, urlToCalloutMap);
+    if (Object.keys(urlToCalloutMap).length == MAX_RTC_CALLOUTS) {
```

Comment:
Nit: have maybeInflateAndAddUrl return boolean of whether it was added so that you only check map length if so.  Also you can get the index in the array from a forEach loop:

rtcConfig['urls'].forEach((url, idx) => ...

Reply:
but you can't force a forEach loop to terminate unless you throw an error, we'd end up looping through the entire array and doing the same check n-5 times. I guess I can change to throw an error, but then I also need to catch that error, seems uglier than using a for-in loop.